Tp objetos - entrega N°1

FedeMacias opened the pull request.

Maikyver committed:
Creación Proyecto

FedeMacias committed:
Creacion del archivo .wlk

FedeMacias committed:
Primera version

una primera version muy basica, vamos modificando en base a esto

Maikyver committed:
Joaquin Modelado

Maikyver committed:
Lucia Modelada

Se debe aun mejorar el interpretaBien para que acepte mayusculas

Maikyver committed:
Luis Alberto Modelado

FedeMacias committed:
Presentaciones

se agregaron las presentaciones

Maikyver committed:
Tests 1-6

FedeMacias committed:
Test 7-13

por alguna razon no los pude correr, pero estoy casi seguro que son asi

Maikyver committed:
Tests pasados

Maikyver committed:
Arreglos para los test 6 a 13

Maikyver committed:
Tests terminados 18-21

Maikyver committed:
Tests 7/14(+-)/15

Maikyver committed:
Tests 22/23

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+	var habilidad = 20  
+	method tuHabilidad() = habilidad
+	
+	method tuHabilidadSiCantasEnGrupo()= habilidad + 5 
```
Debería haber **un** método para responder cuando le preguntan su habilidad. Y que responda 20 + la cantidad que sea en base a si está en un grupo (deleguen a otro método esa parte).

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+	method interpretasBien(cancion) = ( cancion.duracionCancion() > 300 )  
+	
+	method cobras (presentacion) { 
+		 if (presentacion.getsePresentan().size()>1)//Mejorar - Que p�sa si es mayora 1 pero no es el el q esta en la lista
```
La abstracción sería "hay un solo músico" quizá. O bien "se presenta solo... esta persona".

En la primera abstracción, nos fijamos si la cantidad es igual a 1.
En la segunda, recibimos a alguien, y comparamos la lista con **== [elMúsico]**

Elijan lo que gusten, pero que lo haga la **presentación**. Desde acá, solo mandémosle un mensaje. No le revolvamos la colección y le pidamos la cantidad de gente y comparando desde afuera.

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+		 if (presentacion.getsePresentan().size()>1)//Mejorar - Que p�sa si es mayora 1 pero no es el el q esta en la lista
+		return 50
+		else
+		return 100
```
Dos return, en general, es una mala práctica.
Una opción es tener una variable, y en cada rama asignarle un valor, logrando un único return de esa variable al final de todo.

Otra opción, es hacer directamente:
= if (condiciónLoca) valorDeRetorno1 else valorDeRetorno2

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+object joaquin {
+	
+	const grupo = "pimpinela"
```
Con esto, no podemos saber si la persona pertenece a un grupo o a una banda solista. Y vamos a necesitarlo.

Podrían crear algo que represente al grupo pimpinela, y otra cosa que represente a un solista. Y que esas entidades sepan decir si son grupo o no.

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+	
+	method tuHabilidad() = habilidad
+	
+	method tuHabilidadSiCantasEnGrupo() =habilidad - 20
```
Idem de lo de más arriba. Tiene que existir una única manera de preguntarle la habilidad, y que en base a cosas que solo ella conoce, decirnos si es 70 o 70 - 20

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+	
+	method tuHabilidadSiCantasEnGrupo() =habilidad - 20
+	
+	method interpretasBien(cancion) =cancion.letra().contains("familia") //VER COMO ES PARA QUE HACEPTE MAYUSCULAS TAMBIEN
```
¿De quién es la responsabilidad de saber si **su letra** contiene cierta palabra? De la canción. Así que **deleguen**.

Luego, tienen 2 problemas:
- Lo de las mayúsculas que indicaron. Eso se resuelve haciendo .letra().toLowerCase() para que esté todo en minúscula y listo.
- Palabras como "familia**res**". La forma de zafar, es separar el string en una lista de palabras (use "words()"), y luego ver si esa lista contiene exactamente el elemento "familia".

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+	
+	method cobras (presentacion) { 
+		
+		 if (presentacion.lugar().capacidad()>5000)
```
¿Quién debería saber decirnos si es concurrida?

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+ 
+object luisAlberto {
+
+	//var guitarras = [fender, gibson]  
```
Borren lo que no les sirva.

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+
+	//var guitarras = [fender, gibson]  
+
+	method habilidad (guitarra, estado) {
```
Sería conveniente que se guarde una referencia a la única guitarra que esté usando en ese momento.

Y el estado de la guitarra... ¿Quién lo debe saber?

Y la ecuación para saber el precio de una guitarra... ¿Quién es la entidad responsable de saberla?

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+	method interpretasBien(cancion) = true
+	
+	method cobras (presentacion) {
+		if(presentacion.fechaPresentacion() < 20170901)
```
Utilicen objetos instanciados de Date. Son comparables :D

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+	
+	//var valorGuitarra =15
+	method valorGuitarra (estado) {
+		if (estado =="rota")
```
Si el estado es "Rota", esto falla. Está muy acoplado a cómo se escribe a mano un estado, y es propenso a fallas.

Podrían hacer un método "rompete" y "arreglate", por ejemplo, que cambie alguna referencia interna a un **booleano**.

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+
+object laFamilia {
+	method duracionCancion () = 264
+	method letra ()= "Quiero brindar por mi gente sencilla, por el amor brindo por la familia"
```
Cualquier canción va a tener que saber decirnos si tiene cierta palabra en su letra.

Así que van a repetir comportamiento.

Y todas las canciones tienen las mismas referencias y el mismo comportamiento.

Podrían evitar esa repetición :D

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+	var primerPiso = 300
+	method capacidad ()= plantaBaja + primerPiso
+
+	//VER COMO AGREGAR LO DE LOS SABADOS 
```
Esto depende de la fecha en la cual consulten la capacidad.

Desde afuera, queremos decirle: "Che, qué capacidad tenés para el 3/9/17?". Así que el método capacidad, tiene que recibir la **fecha**.

¿Y qué pasa con el lunaPark? **También** tiene que recibir la fecha, por más que no la use, así entiende el mismo mensaje y los podemos tratar polimórficamente.

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,122 @@
+		method fechaPresentacion () =20170420
+	}
+	object presentacionTrastienda{
+		var sePresentan = [luisAlberto, joaquin, lucia]
```
Las presentaciones tienen todas la misma estructura, y con los cambios de antes, van a compartir algunos comportamientos.

Siendo así, conviene evitar repetir de alguna forma.

Recuerden que pueden crear sus objetos a partir de ...complete aquí el concepto... directamente en el fixture. No hace falta crear cada entidad en este archivo, sino alguna forma de poder crearlas en un programa o en un test. ¿Me explico?

sanuzzi commented on test.wtest:
```diff
@@ -0,0 +1,116 @@
+import TpObjetos.*
+test "1.Joaqu�n canta en el grupo Pimpinela. Queremos conocer su habilidad" {
```
El título del test no está bueno. Recién se entiende qué esperamos leyendo el cuerpo.
Traten de mejorar esto en todos los tests.

Maikyver replied:
copié los test como estaban en la consigna

sanuzzi commented on test.wtest:
```diff
@@ -0,0 +1,116 @@
+}
+test "2.Joaqu�n canta en el grupo Pimpinela, y decide lanzar su carrera solista. Queremos conocer su habilidad" {
+
+	assert.equals(joaquin.tuHabilidad(), 20)
```
Hay que indicarle que su banda, agrupación o como quieran decirle, es un objeto "solista" en vez de "pimpinela". Y que **no** sean strings sino objetos que nos sepan decir si son grupo o no.

Y luego preguntarle la habilidad.

sanuzzi commented on test.wtest:
```diff
@@ -0,0 +1,116 @@
+}
+test "3.Lucia canta en el grupo Pimpinela. Queremos conocer su habilidad" {
+
+	assert.equals(lucia.tuHabilidadSiCantasEnGrupo(), 50)
```
El método para la habilidad debe ser idéntico al de antes.
Pero al estar en pimpinela, nos debe decir 50 él solito.

sanuzzi commented on test.wtest:
```diff
@@ -0,0 +1,116 @@
+}
+test "5.Luis Alberto toca con la Fender. Queremos saber su habilidad" {
+
+	var sana//porque....
```
Eh? var sana//porque.... ?

Bueno, ya que hablamos de Luis: Tiene la guitarra fender, y está sana, así que podrían no poner nada de eso y asumir que siempre empezará así (ok, sería asumir cosas de más quizá, pero no me hace mucho ruido)

Maikyver replied:
El test no me funcionaba a menos que le ponga el var sana. Tenia que averiguar porque era eso y me olvide :D

sanuzzi commented on test.wtest:
```diff
@@ -0,0 +1,116 @@
+}
+test "6.Luis Alberto toca con la Guibson sana. Queremos saber su habilidad" {
+
+	var sana//porque....
```
Acá tenemos que decirle a la guitarra gibson que se rompa.

Y a Luis, que ahora toque con esa gibson.

Abajo le preguntamos la habilidad con **el mismo** mensaje que a los otros músicos, y que él nos diga.

sanuzzi commented on test.wtest:
```diff
@@ -0,0 +1,116 @@
+}
+test "9.Queremos saber si Joaqu�n interpreta bien la canci�n �Cisne�" {
+	
+	var cisne
```
Las canciones deberían estar creadas en el fixture, así las podemos reutilizar en varios tests

sanuzzi commented on test.wtest:
```diff
@@ -0,0 +1,116 @@
+}
+test "22.Queremos saber el costo de la presentaci�n del Luna Park"
+ {
+ 	assert.equals(presentacionLunaPark.getsePresentan().sum({persona => persona.cobras(presentacionLunaPark)}),1550)
```
No!!! Desde afuera, **no** le revolvamos la colección. Que la presentación se encargue de decirnos su costo. Deleguen a ella la responsabilidad de eso.

sanuzzi reviewed (changes requested):
Primera revisión de la entrega 1. Partan de estos comentarios y revisen todo el código por más mejoras.

Ojo: Si editan el código, el comentario desaparece. Pueden descargarse la página por las dudas, o asegurarse de resolverlo completamente antes de hacer un commit.

Maikyver committed:
Cancion e intento de Musico

Maikyver committed:
Error- Help

porque no me los sobrescribe??

Maikyver committed:
Presentacion/ LunaPark/LaTrastienda

Hay que arreglar porq en presentacion se le manda un objeto lugar q es LunaPark  o Larastienda pero nos e como hacerlo desde el programa eso

Maikyver committed:
Cancion+Album

Utilice un constructor para "setearla" 
Creacion de Album

FedeMacias committed:
Agregadas Nuevas Funcionalidades

FedeMacias committed:
Modificado vocalita popular

falta hacer lo de la guitarra rota y la sana

Maikyver committed:
Arrelgos a Musico y Album

Maikyver committed:
Merge remote-tracking branch 'origin/develop' into develop

Maikyver committed:
Guitarras

Maikyver committed:
Pegó

Ahora depende del album y no el musico q me diga si fue buena venta

Maikyver committed:
Areglo de sos minimalista

arreglo de canción para saber si es corta

FedeMacias committed:
Fixture casi completo

Todo lo que este en mayusculas son anotaciones

Maikyver committed:
me duele la cabeza...

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+	var habilidad = 0
+	var albumes = []
+	
+	constructor (tipoDeMusicoM,habilidadM,albumesM)
```
la "M" trae confusión

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+
+
+ 
+/*object luisAlberto  {
```
Hacer que Luis Alberto sea una instancia de Músico en el fixture, y que su tipo sea el **objeto** tipoDeMúsicoDeLuisAlberto con sus cosas especiales

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+}
+
+class VocalistaPopular {
+	var interpretasBienPalabra 
```
Usen mejores nombres para la palabra que se guardan para interpretar bien

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+ 
+/*object luisAlberto  {
+
+	var guitarras = [fender, gibson]  
```
Que se guarde una sola guitarra, la que usa **ahora**.

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+		self.albumesPublicados(albumesP)
+		
+	} 
+	method habilidad (guitarra, estado) {
```
Que el estado lo tenga la guitarra, y que ella sepa decirnos su valor

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+		
+	} 
+	method habilidad (guitarra, estado) {
+		if (guitarra.valorGuitarra(estado) * 8<100) 
```
"valorGuitarra" --> "valor"

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+		
+			return guitarra.valorGuitarra(estado) * 8
+		else
+			return 100
```
Vean la guía de lenguajes --> "min" (para números)

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+	override method interpretasBienLaCancion(cancion) = true
+	
+	override method cobras (presentacion) {
+		if(presentacion.fechaPresentacion() < 20170901)
```
= if (condición) valor1 else valor2

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+ 
+
+class Cancion{
+	var tituloPrincipal
```
pongan "título"

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+	var duracion
+	constructor (tituloPrincipalAlbum,letraCancion, duracionCancion)
+	{
+		letra = letraCancion
```
Si esperan que sea constante y nunca cambie, está bien que no esté el setter, pero hagan las referencias const.

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+	method letra() = letra	
+	method duracionCancion ()= duracion 
+	method contienePalabra(palabra){
+	 	return self.letra().words().contains(palabra)
```
.letra().toLowerCase()....

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+	method unidadesALaVenta()=unidadesALaVenta
+	method unidadesVendidas()=unidadesVendidas
+	method duracion()=self.canciones().sum({cancion => cancion.duracionCancion()})
+	method cancionMasLarga()=self.canciones().max({cancion => cancion.letra().lenght()})
```
Que la canción sea la responsable de saber la extensión de su letra

sanuzzi commented on TpObjetos.wlk:
```diff
@@ -0,0 +1,183 @@
+
+object laTrastienda {
+	var plantaBaja = 400
+	var primerPiso = 300
```
Podrían ser métodos que devuelvan eso valores

sanuzzi commented on test.wtest:
```diff
@@ -0,0 +1,254 @@
+	assert.equals(lucia.tuHabilidadSiCantasEnGrupo(), 50)
+
+}
+test "4.Lucia canta en el grupo Pimpinela, y decide lanzar su carrera solista. Queremos conocer su habilidad" {
```
Traten de mejorar los nombres de los tests

Maikyver committed:
Clases separadas por archivos

Tambien creado test entrega dos, para ser más cómodo

Maikyver committed:
Primeros 4 test bien hechos

FedeMacias committed:
Test de 5 al 16

no funcionan bien el   7,11,12 y 13 , mirar el de si la pegó soledad o no

FedeMacias committed:
No importa este cambio

Maikyver committed:
Los tipos de musico son objetos y test 11-12-13

Maikyver committed:
Entrega 2

No deja hacer un nuevo pull request si no se acepta el anterior.

Maikyver committed:
Test 1

EL testt esta bien hecho pero el del tp de otra cosa por algun requerimiento q no especificaron

Maikyver committed:
Tests del 2 al 7

El test 5 también dio distinto que el test de la entrega pero tambien especifiqué porque es que nos da bien.

FedeMacias committed:
Funcion Remixeate

FedeMacias committed:
Version primaria de comparar 2 canciones

acordarse de crear "otraCancion" como new cancion  en el fixture

FedeMacias committed:
Este cambio no importa

Maikyver committed:
Remix- Test 4 y 5

Cambie el Remix para que devuelva una cancion distinta a la original. De esta manera, no se toca la original y tampoco se debe llamar cada vez que se desee hacer algo con el remix.

Maikyver committed:
Mashup

Tambien cambie el comparar para q sea con max y no con ifs

FedeMacias committed:
Otro cambio no importante

FedeMacias committed:
Pdpalooza (no está 100% funcional)

no me reconoce el "throwException" entonces falla, y dudo sobre si conviene inicializar la cancion directamente en la condicion para poder presentarse

FedeMacias committed:
Pdpalooza completo

funcionan los exception e inicializo la cancion en la funcion para ver si cumple las condiciones

Maikyver committed:
Test 7 y 8

Arreglo en pdpalooza para que funcione

Maikyver committed:
Test 9 a 11

arregle lña exception de la "Cancion de Alicia en el País" (siempre iba a cantarla bien porque el requerimiento apra cantar bien ahora es que si tiene mas de 60 de habilidad canta bien todo. por loq se pregunta al tipo, no al metodo interpreta bien)

Maikyver committed:
Test 11...

yo tenia razon con soledad.... se reutiliza el method y ahora si se puede agregar a pdpalooza

sanuzzi commented on Album.wlk:
```diff
@@ -0,0 +1,27 @@
+	method unidadesALaVenta()=unidadesALaVenta
+	method unidadesVendidas()=unidadesVendidas
+	method duracion()=self.canciones().sum({cancion => cancion.duracionCancion()})
+	method cancionMasLarga()=self.canciones().max({cancion => cancion.letra().words().size()})
```
cancion.letra().size() --> Que lo sepa la canción.

Y el criterio usado en max, debe recibirse por parámetro en:
canciónMásLargaSegún(criterio) = self.canciones().max( {cancion => criterio.tamaño(cancion) } )

Y esto queda:
method cancionMasLarga() = self.canciónMásLargaSegún( tamañoSegunLetra )

sanuzzi commented on Cancion.wlk:
```diff
@@ -0,0 +1,46 @@
+	}
+
+
+	 method comparaTuDuracionCon(otraCancion) =self.duracionCancion().max(otraCancion.duracionCancion())
```
Estos 3 se van.
Sí podría ser necesario un "extensiónDelTítulo" y "extensiónDeLaLetra"

sanuzzi commented on Cancion.wlk:
```diff
@@ -0,0 +1,46 @@
+	 
+	 method mashupeateCon(unasCanciones){
+	 	const unacancion =new Cancion(self.titulo() , self.letra(),self.duracionCancion())
+	 	unasCanciones.forEach({cancion =>unacancion.letra(cancion.letra() +" "+ unacancion.letra() )})
```
Hagan un map, con trim y join(" "), para letra y para título, y traten de repetir lo menos posible

sanuzzi commented on Cancion.wlk:
```diff
@@ -0,0 +1,46 @@
+	 method sosCorta ()= self.duracionCancion()<180 
+	 
+	 
+	 method remixeate(){
```
Se podría hacer una clase Remix y una Mashup.

Cada una tendrá **su** constructor con lo que necesite.

class Remix inherits Cancion {
	constructor(unaCancion) = super(
		unaCancion.titulo(),
		unaCancion.duracion() * 3,
		 "sadsa" + unaCancion.letra() + "asdasd"  ) {}
}

class Mashup inherits Cancion {
	constructor(canciones) {
		letra = self.concatenarSegun({cancion => cancion.letra()}, canciones)
		titulo = self.concatenarSegun({cancion => cancion.titulo()}, canciones)
		duracion = tamanioPorDuracion.elMasGrande(canciones).duracion()
	}
	method concatenarSegun(criterio, canciones) = 
		canciones.map({cancion => criterio.apply(cancion).trim()}).join(" ")
}

// Volviendo a los criterios pa' comparar:

class CriterioParaCompararCanciones {
	method elMasGrande(unasCanciones) = unasCanciones.max({unaCancion => self.tamanio(unaCancion)})	
	method tamanio()
}
object tamanioPorDuracion inherits CriterioParaCompararCanciones {
	method tamanio(unaCancion) = unaCancion.duracion()
}
... y así para los 3 criterios

sanuzzi commented on Cancion.wlk:
```diff
@@ -0,0 +1,46 @@
+class Cancion{
+	var tituloPrincipal
```
¿Por qué variable?

sanuzzi commented on Presentacion.wlk:
```diff
@@ -0,0 +1,43 @@
+		}
+	}
+	}
+object pdpalooza inherits Presentacion{
```
1. Se puede agregar un músico y no se valida nada. Hacer que se valide. Entonces, hay 2 override.
2. Hacer una clase PresentaciónConRestricciones.
3. Que las restricciones las conozcan los objetos instanciados de esta clase.
4. Las restricciones sabrán cuando mucho 2 cosas: Si rechazan a alguien, y el texto descriptivo si no se cumple.
5. Usen una nueva clase de excepción.
6. Al agregar muchos músicos, verifiquen esto mismo para todos.

sanuzzi commented on test.wtest:
```diff
@@ -0,0 +1,198 @@
+	presentacionLunaPark.fechaPresentacion (20,04,2017)
+	
+	presentacionLaTrastienda = new Presentacion()
+	//presentacionLaTrastienda.sePresentan(luisAlberto, joaquin, lucia)
```
Quiten comentarios

sanuzzi commented on test.wtest:
```diff
@@ -0,0 +1,198 @@
+
+	albumParaLosArboles = new Album("Para los Arboles",[cisne,almaDeDiamante],[31,03,2003],50000,49000)
+
+	albumJustCrisantemo = new Album("Just Crisantemo",[crisantemo],[05,12,2007],28000,27500)
```
Que las fechas, sean fechas!! new Date(día, mes, año)

Que los **tres archivos de tests** queden funcionando

sanuzzi commented on Entrega3.wtest:
```diff
@@ -0,0 +1,172 @@
+}
+test "1.Joaqu�n NO canta bien la canci�n 'La Familia'(su habilidad es 25 por ser de grupo, solo tiene un album y la cancion dura menos de 300(264) " {
+
+	assert.notThat(joaquin.interpretasBien(laFamilia))
```
// Acá le settean los álbumes a artistas y demás
// Podría estar esta línea en el fixture al principio:
fixtureAuxiliar.inicializate()

const joaquin = fixtureAuxiliar.joaquin()
const laFamilia = fixtureAuxiliar. laFamilia()

assert.....

sanuzzi commented on Entrega3.wtest:
```diff
@@ -0,0 +1,172 @@
+
+}
+test "3.Luc�a canta bien la cancion 'La Familia' como solista" {
+ 	lucia.tipoDeMusico(vocalistaPopular)
```
Si siempre o generalmente es vocalista popular, que esté en el fixture o en el objeto auxiliar

sanuzzi commented on Entrega3.wtest:
```diff
@@ -0,0 +1,172 @@
+test "10. Lucia como solista NO se puede agregar al recital PDpalooza" {
+ 	lucia.tipoDeMusico(vocalistaPopular)
+	pdpalooza.sePresentan([lucia])
+	assert.that(true)
```
Incompleto. Revisen apuntes.

sanuzzi commented on TestsEntrega2.wtest:
```diff
@@ -0,0 +1,150 @@
+	//lucia.intrepretasBien("familia")
+	
+	soledad = new Musico(vocalistaPopular,55,[albumLaSole])
+	//soledad.interpretasBien( "amor")
```
Turbio!! Sáquenlo!!

sanuzzi commented on program.wpgm:
```diff
@@ -0,0 +1,86 @@
+		
+	joaquin = new Musico(new MusicoDeGrupo(),20,[albumLaFamilia])
+	
+	lucia = new Musico(new VocalistaPopular(),70,[])
```
Esto se rompe. No hay una clase VocalistaPopular. Arreglen lo anterior.

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,129 @@
+	method cancionMasLarga() = self.albumesPublicados().filter({album => album.cancionMasLarga()})     
+
+	method laPegaste() = self.albumesPublicados().all({album => album.buenaVenta()})
+	method palabraInterpretaBien(unaPalabra){self.tipoDeMusico().palabra(unaPalabra)}
```
Se puede inicializar al tipo desde afuera. Ese cacho de código no va, porque con el tipo de grupo se rompe.

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,129 @@
+	 	}
+	}
+
+object vocalistaPopular {
```
Debe ser clase, para que cada uno tenga su palabra.

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,129 @@
+	method interpretasBien(cancion) = cancion.contienePalabra(self.palabra())
+} 
+object musicoDeGrupo{
+	method tuHabilidad(habilidad) = habilidad + 5 
```
También tiene que tener un constructor, para los puntos adicionales de la habilidad

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,129 @@
+	method interpretasBien(cancion) { return cancion.duracionCancion()>300}
+}
+
+object luisAlberto {
```
Mejoren el nombre, como tipoDeMúsicoDeLuisAlberto, o una clase MúsicoReCapo.

sanuzzi reviewed (commented):
Revisión entrega 3

FedeMacias committed:
Ningun cambio mayor

Deje algo masomenos preparado para encararlo despues

FedeMacias committed:
Otros cambios menores

FedeMacias committed:
Entrega 3 al dia

Maikyver committed:
Test 1

todos listos menos los q sean relacionados con cobrar

FedeMacias committed:
Nada importante

FedeMacias committed:
Hecho palabrero, larguero e imparero

Tambien se agrego metodo esDeTuAutoria()

FedeMacias committed:
Hechas banda y representante

Faltaria todo lo que tiene que ver con lo que cobran los artistas y las bandas. Y ademas faltan los requerimientos adicionales y los tests

FedeMacias committed:
Idea inicial de TipoDeCobro

Lo que está en comentarios es todo lo que pense de como hacer el TipoDeCobro

FedeMacias committed:
habilidadBanda(), cualesInterpretasBien(), magia()

Maikyver committed:
Tests Entrega 3 y 4

Aregglo primeros dos de la entrega 3 y hechos los primeros 6 de la entrega 4

Maikyver committed:
Creados los tipo de cobro

Tipos de crobros. Reforme presentacion. Test del 5 al 11

FedeMacias committed:
habilidad polimorfismo

FedeMacias committed:
Test 12 a 17

Maikyver committed:
Entrega 4

Test del 12 al 18 con test positivo.

FedeMacias committed:
Intente hacer los diagramas

sanuzzi commented on PresentacionConRestricciones.wlk:
```diff
@@ -0,0 +1,27 @@
+import Musico.*
+import Cancion.*
+
+class PresentacionConRestricciones inherits Presentacion {
```
Hace mucho ruido que esta clase no aporte nada de comportamiento útil.

Acá podrían tener una colección de restricciones, y validar si se cumplen.

Luego, podrían tener clases de restricciones parametrizables para armar lo que quieran.

sanuzzi commented on PresentacionConRestricciones.wlk:
```diff
@@ -0,0 +1,27 @@
+class PresentacionConRestricciones inherits Presentacion {
+method puedePresentarse(unaPersona){}
+}
+object pdpalooza inherits PresentacionConRestricciones{
```
Entonces, pDePalooza sería un objeto **instaciado** de la clase, y no un WKO.

sanuzzi commented on PresentacionConRestricciones.wlk:
```diff
@@ -0,0 +1,27 @@
+		const cancion = new Cancion("cancion de alicia en el pais","Qui�n sabe Alicia, este pa�s no estuvo hecho porque s�. Te vas a ir, vas a salir pero te quedas, �d�nde m�s vas a ir? Y es que aqu�, sabes el trabalenguas, trabalenguas, el asesino te asesina, y es mucho para ti. Se acab� ese juego que te hac�a feliz.",510)
+	
+		if(unaPersona.esHabilidoso().negate()){
+			throw new Exception("el musico no puede participar si posee menos de 70 puntos de habilidad")
```
Que la lógica de lanzar la excepción esté una sola vez.

Que el mensaje esté en cada restricción.

Usen una clase personalizada.

sanuzzi commented on Representante.wlk:
```diff
@@ -0,0 +1,6 @@
+	method cobras(bandaCobra)=1982
+}
+object felix{
+	method cobras(bandaCobra)=bandaCobra*0.1
```
Puede ser una clase y crear los representantes que queramos

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,181 @@
+	method cobras(presentacion){if (presentacion.fecha()>self.fecha()) return self.cobro()*(1+self.porcentajeAdicional()) else return self.cobro()}
+}
+
+class MusicoComoLuisAlberto {
```
Hagan que luisAlberto sea un WKO
Pero que **no** herede de Músico.
Que herede de MúsicoBásico / MúsicoAbstracto / MúsicoGenérico / Músico (y que el "Músico" que tenían se llame de otra forma).

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,181 @@
+class Musico {
+
+	var tipoDeMusico 
+	var habilidad = 0
```
Faltaría agregar acá lo de conocer al objeto para saber si interpreta bien la canción

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,181 @@
+	method interpretasBien(cancion) = true
+	
+	method cobras (presentacion) {
+		const fecha = new Date(01,09,2017)
```
= if(presentacion.fecha() < new Date(01,09,2017)) 1000 else 1200

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,181 @@
+object gibson{
+	 
+	 var estadoRota = false
+	method valorGuitarra () {
```
= if (...) 5 else 15

sanuzzi commented on Presentacion.wlk:
```diff
@@ -0,0 +1,49 @@
+import Banda.*
+	class Presentacion{
+		
+		var sePresentan =[]
```
Replanteen cuándo usar lista y cuándo set

sanuzzi commented on Presentacion.wlk:
```diff
@@ -0,0 +1,49 @@
+	method plantaBaja() = 400
+	method primerPiso() = 300
+	method capacidad (fecha){
+		if(fecha.dayOfWeek()!=6) 
```
capacidad(fecha) = self.capacidadEnPlantaBaja() + capacidadExtraSiEsSabado(fecha)

sanuzzi commented on PresentacionConRestricciones.wlk:
```diff
@@ -0,0 +1,27 @@
+}
+object pdpalooza inherits PresentacionConRestricciones{
+	override method sePresentan(nuevasPersonas){
+		nuevasPersonas.forEach({unaPersona => self.puedePresentarse(unaPersona)
```
self.lanzarExcepcionSiAlgunoNoPuedeEntrar(nuevasPersonas)
self.agregaMusicos(nuevasPersonas)

method lanzarExcepcionSiAlgunoNoPuedeEntrar(nuevasPersonas) =
nuevasPersonas.forEach({persona => self.lanzarExcepcionSiNoPuedeEntrar(unaPersona)})

method lanzarExcepcionSiNoPuedeEntrar(unMusico){
    self.restricciones().forEach({restriccion => restriccion.rompeteSiNoLoDejasPasar(unMusico)})
}

La superclase Restricción tendría lo de rompeteSiNoLoDejasPasar, y delegaría lo de la condición y el mensaje.

Las subclases tendrían esos 2 métodos implementados.

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,181 @@
+
+class Musico {
+
+	var tipoDeMusico 
```
Que el tipo de músico siga siendo VocalistaPopular vs. MúsicoDeGrupo, y que esto resuelva lo de la habilidad

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,181 @@
+	}
+	method cobro()=cobroSolo
+	method cobro(nuevoCobroSolo){cobroSolo = nuevoCobroSolo}
+	method cobras(presentacion){if (presentacion.sePresentan().equals([self])) return self.cobro() else return self.cobro()/2}
```
No le manoseen la colección. Que la presentación nos diga si el único que se presenta es ese flaco.

Y ojo, no es self, es el chabón

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,181 @@
+}
+
+class CobroSegunCapacidad{
+	var cobro
```
Si quieren, que sean constantes, pensando en el caso de uso: Che, tu nueva forma de cobrar es new Saraza

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,181 @@
+	constructor(unCobro,unaCantPersonas)
+	{
+		cobro=unCobro
+		cantPersonas=unaCantPersonas
```
Si lo dejan variable, ya que tienen un setter, pueden usarlo

sanuzzi commented on Musico.wlk:
```diff
@@ -0,0 +1,181 @@
+	method cobro(nuevoCobro){cobro = nuevoCobro}
+	method cantPersonas()=cantPersonas
+	method cantPersonas(unaCantPersonas){cantPersonas = unaCantPersonas}
+	method cobras(presentacion){if (presentacion.lugar().capacidad()>self.cantPersonas()) return self.cobro() else return self.cobro()-100}
```
Deleguen lo de saber la cantidad de gente que va a haber en la presentación

sanuzzi commented on Banda.wlk:
```diff
@@ -0,0 +1,23 @@
+	method tusIntegrantes()= integrantes
+	method tuRepresentante()= representante
+	method tuNombre()= nombre
+	method tuHabilidad() = self.tusIntegrantes().sum({musico => musico.tuHabilidad()})  * 1.1
```
* plusPorQuímica() o como lo quieran hacer

sanuzzi commented on Banda.wlk:
```diff
@@ -0,0 +1,23 @@
+	method tuNombre()= nombre
+	method tuHabilidad() = self.tusIntegrantes().sum({musico => musico.tuHabilidad()})  * 1.1
+	method bandaCobra(presentacion)=self.tusIntegrantes().sum({musico => musico.cobras(presentacion)})
+	method cobras(presentacion) = self.bandaCobra(presentacion) + self.tuRepresentante().cobras(self.bandaCobra(presentacion))
```
Se podría hacer una lista de la gente relacionada a la banda (con músicos y representantes), que se llame "todos", "todosLosRelacionados", "todosLosQueCobran", "integrantes", o como gusten.

method todos() = self.musicos() + [self.representante()]

method cobras(presentacion) = self.todos().sum({alguienQueCobra => alguienQueCobra.cobras(presentacion)})

Maikyver committed:
Arreglos de la Entrega 4

Tambien arrelgados los tests de todas las entregas

Maikyver committed:
Entrega Final

Todo el código, test y diagrama de Clases
